lib/utils/FileCache.mjs

```diff
@@ -60,23 +60,54 @@ export default class FileCache {
 		// Get file descriptor compares current metadata against cached
 		// one and stores the result to "changed" prop.w
 		const descriptor = this._fileCache.getFileDescriptor(absoluteFilepath);
-		/** @type {CacheMetadata} */
-		const meta = descriptor.meta || {};
-		const changed = descriptor.changed || meta.hashOfConfig !== this._hashOfConfig;
+
+		descriptor.meta.data ??= {};
+
+		assert(isPlainObject(descriptor.meta.data));
+
+		const configChanged = descriptor.meta.data.hashOfConfig !== this._hashOfConfig;
+
+		let changed = false;
+
+		if (this._useCheckSum) {
+			changed = configChanged || this.hasFileCheckSumChanged(absoluteFilepath);
+		} else {
+			changed = configChanged || Boolean(descriptor.changed);
+		}
 
 		if (!changed) {
 			debug(`Skip linting ${absoluteFilepath}. File hasn't changed.`);
 		}
 
 		// Mutate file descriptor object and store config hash to each file.
 		// Running lint with different config should invalidate the cache.
-		if (meta.hashOfConfig !== this._hashOfConfig) {
-			meta.hashOfConfig = this._hashOfConfig;
+		if (descriptor.meta.data.hashOfConfig !== this._hashOfConfig) {
+			descriptor.meta.data.hashOfConfig = this._hashOfConfig;
 		}
 
 		return changed;
 	}
 
+	/**
+	 * @param {string} absoluteFilepath
+	 * @returns {boolean}
+	 */
+	hasFileCheckSumChanged(absoluteFilepath) {
+		const metaCache = this._fileCache.cache.getKey(this._fileCache.createFileKey(absoluteFilepath));
+
+		if (!metaCache?.hash) {
+			return true;
+		}
+
+		const buffer = readFileSync(absoluteFilepath);
+
+		if (this._fileCache.getHash(buffer) !== metaCache.hash) {
+			return true;
+		}
+
+		return false;
+	}
```

Comment:
This is based on : https://github.com/jaredwray/cacheable/blob/5ec1858bdd07aa5efc30e87f14253a951fbc83b5/packages/file-entry-cache/src/index.ts#L243-L310